File: python_src/cluster_independent.py
from optparse import Option
import pandas as pd
import scanpy as sc
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import anndata

from scipy.spatial import distance_matrix
from scipy.stats import wasserstein_distance
from statsmodels.nonparametric.kernel_density import KDEMultivariate
from scipy.stats import entropy

from multiprocessing import Pool
from typing import Optional, Iterable

def rank_by_proportion_within_threhold(
    adata: anndata,
    thresholds: Optional[Iterable | int] = np.linspace(1, 501, 21),
    guide_list: Optional[Iterable | None] = None,
    control_guide: Optional[str] = 'sgNon-targeting',
    method: Optional[str] = 'bin',
    kernel_func: Optional[str] = 'log',
    kernel_scoring: Optional[Iterable | None] = None,
    return_fig: Optional[bool] = False
) -> None:
    proportion = {}
    if type(thresholds) is int:
        thresholds = np.linspace(1, thresholds, 21)
    if guide_list == None:
        guide_list = adata.var_names.tolist()
    if method == 'bin':
        for threshold in thresholds:
            proportion[threshold] = []
            for guide in guide_list:
                if guide == control_guide: continue
                guide_data = adata[adata[:, guide].X > 0]
                ntc_data = adata[adata[:, control_guide].X > 0]
                in_distance = ((distance_matrix(guide_data.obsm['spatial'], ntc_data.obsm['spatial']) < threshold).sum(axis=1) > 0).sum()
                proportion[threshold].append(in_distance / guide_data.shape[0])
    elif method == 'count':
        for threshold in thresholds:
            proportion[threshold] = []
            for guide in guide_list:
                if guide == control_guide: continue
                guide_data = adata[adata[:, guide].X > 0]
                ntc_data = adata[adata[:, control_guide].X > 0]
                in_distance = guide_data[(distance_matrix(guide_data.obsm['spatial'], ntc_data.obsm['spatial']) < threshold).sum(axis=1) > 0, guide].X.sum()
                proportion[threshold].append(in_distance / guide_data.shape[0])
    else:
        print('Error, method must be one of "bin" or "count"!')

    d_df = pd.DataFrame(proportion).melt()
    d_df.index = adata[:, (adata.var_names != control_guide) & np.isin(adata.var_names, guide_list)].var_names.tolist() * len(thresholds)
    d_df.reset_index(inplace=True)

    if kernel_scoring != None:
        sc = d_df.apply(lambda x: np.multiply(kernel_scoring, 1 - x), axis=1).sum(axis=1).sort_values()
    else:
        if kernel_func == 'log': 
            ln_scoring = np.array([np.log2(x) for x in np.linspace(1, np.exp2(10), len(thresholds))])
            sc = d_df.apply(lambda x: np.multiply(ln_scoring, 1 - x), axis=1).sum(axis=1).sort_values()
        elif kernel_func == 'linear':
            linear_scoring = np.linspace(0, 10, len(thresholds))
            sc = d_df.apply(lambda x: np.multiply(linear_scoring, 1 - x), axis=1).sum(axis=1).sort_values()
        elif kernel_func == 'exp':
            exp_scoring = np.array([np.power(2, x) for x in np.linspace(1e-32, np.log2(10), len(thresholds))])
            sc = d_df.apply(lambda x: np.multiply(exp_scoring, 1 - x), axis=1).sum(axis=1).sort_values()

    df = pd.DataFrame(sc, columns=['dist'])
    df['guide'] = df.index.str.split('_').str.get(0).tolist()
    df.reset_index(inplace=True)
    _, ax = plt.subplots(1, 1, figsize=(len(guide_list) // 3 * 10, 6))
    sns.barplot(df, y='ln_dist', x='guide', ax=ax, palette='rainbow', orient='v', alpha=0.6)
    plt.xticks(rotation=90)
    
    plt.gca().set_yscale('log')
    plt.xlabel('Gene')
    plt.ylabel('Distance score')
    plt.show()

    if return_fig: return ax
    else: return

def rank_by_kernel_estimated_distance(
    adata: anndata,
    control_guide: Optional[str] = 'sgNon-targeting',
    guide_list: Optional[Iterable] = None,
    n_permutation: Optional[int] = 50,
    n_process: Optional[int] = 8,
    return_fig: Optional[bool] = False,
    return_dataframe: Optional[bool] = True,
    sort_by_replicate: Optional[str | None] = '_'
) -> None:
    
    def compute_kde_statsmodels(data):
        kde = KDEMultivariate(data, var_type='ccu')
        return kde.pdf(data)

    def kde_worker(args):
        guide, control, spatial_coords, data, n_permutation = args
        if guide == control: return
        gene_a_expression = data[:, guide].X.flatten()
        gene_b_expression = data[:, control].X.flatten()
        data_a = np.vstack([spatial_coords[:, 0], spatial_coords[:, 1], gene_a_expression]).T
        data_b = np.vstack([spatial_coords[:, 0], spatial_coords[:, 1], gene_b_expression]).T
        Za = compute_kde_statsmodels(data_a)
        Zb = compute_kde_statsmodels(data_b)
        actual_distance = wasserstein_distance(Za, Zb)
        combined_expression = np.concatenate([gene_a_expression, gene_b_expression])
        permuted_distances = []
        np.random.seed(42)
        for _ in range(n_permutation):
            np.random.shuffle(combined_expression)
            permuted_a = combined_expression[:len(gene_a_expression)]
            permuted_b = combined_expression[len(gene_a_expression):]
            permuted_data_a = np.vstack([spatial_coords[:, 0], spatial_coords[:, 1], permuted_a]).T
            permuted_data_b = np.vstack([spatial_coords[:, 0], spatial_coords[:, 1], permuted_b]).T
            permuted_Za = compute_kde_statsmodels(permuted_data_a)
            permuted_Zb = compute_kde_statsmodels(permuted_data_b)
            permuted_distance = wasserstein_distance(permuted_Za, permuted_Zb)
            permuted_distances.append(permuted_distance)
        p_value = np.sum(np.array(permuted_distances) >= actual_distance) / len(permuted_distances)
        return guide, actual_distance, p_value
    
    if guide_list == None:
        guide_list = adata.var_names.tolist()
    if return_fig and return_dataframe:
        print('Error, can only return one of fig or dataframe!')
        return
    filtered_data = adata[adata[:, guide_list].X.sum(axis=1) > 0]
    spatial_coords = filtered_data.obsm['spatial']
    with Pool(processes=n_process) as pool:
        results = pool.map(kde_worker, [(guide, control_guide, spatial_coords, filtered_data, n_permutation) for guide in guide_list])
        d_df = pd.DataFrame(filter(None, results), columns=['guide', 'wd', 'p_value']).sort_values(by='wd', ascending=True)

    _, ax = plt.subplots(1, 1, figsize=(len(guide_list) // 4, 8))

    if sort_by_replicate is not None:
        d_df['replicate'] = d_df['guide'].str.split(sort_by_replicate).str.get(0)
        positions =np.arange(len(d_df)) * 2
        width = 0.5

        ax.bar(positions - width/2, np.array(d_df.iloc[:, -2]).flatten(), width=width, label='1')
        ax.bar(positions + width/2, np.array(d_df.iloc[:, -1]).flatten(), width=width, label='2')

        for idx, name in enumerate(d_df.index):
            ax.hlines(y=max([d_df.iloc[idx]['-log(wd)_1'], d_df.iloc[idx]['-log(wd)_2']]) + 0.02, xmin=2*idx-0.5, xmax=2*idx+0.5, colors='black')
            ax.text(x=2*idx, y=max([d_df.iloc[idx]['-log(wd)_1'], d_df.iloc[idx]['-log(wd)_2']]) + 0.03, s='*', ha='center')

        ax.set_xticks(positions)
        ax.set_xticklabels(d_df.index, rotation=90)
        ax.set_xlabel('Guide(Ordered by min in each pair)')
        ax.set_ylabel('-log(WD)')
        ax.set_ylim(np.min(d_df.iloc[:, -2:]) - 0.1, np.max(d_df.iloc[:, -2:]) + 0.1)
        plt.show()
    else:
        sns.barplot(d_df, x='guide', y='wd', hue='guide', legend=False, palette='rainbow')

        for idx, name in enumerate(d_df.index):
            if (d_df.loc[name, 'wd'] > 0) and (d_df.loc[name, 'wd'] > 0):
                ax.hlines(y=d_df.iloc[idx]['wd'] + 0.02, xmin=idx-0.25, xmax=idx+0.25, colors='black')
                ax.text(x=idx, y=d_df.iloc[idx]['wd'] + 0.03, s=('*' if d_df.p_value < 0.05 else '.'), ha='center')

        ax.set_xticklabels(d_df.index, rotation=90)
        ax.set_xlabel('Guide')
        ax.set_ylabel('WD')
        ax.set_ylim(np.min(d_df.loc[d_df.wd > 0, 'wd']) - 0.1, np.max(d_df.wd) + 0.1)
        plt.show()

    if return_dataframe: return d_df
    if return_fig: return ax
    return

def rank_by_relative_entropy(
    adata: anndata,
    reference_guide: Optional[str] = 'sum',
    control_guide: Optional[str] = 'sgNon-targeting',
    result_field: Optional[str] = 'KL distance',
    guide_list: Optional[Iterable] = None,
    n_top: Optional[int] = 50
) -> None:
    entro = {}
    if reference_guide == 'sum':
        qk_guide = adata.X.sum(axis=1).flatten() / adata.X.sum()
    elif reference_guide == 'ntc':
        qk_guide = adata[:, control_guide].X.flatten() / adata[:, control_guide].X.sum()
    else:
        print('Error, reference_guide must be one of "sum" or "ntc"!')
        return
    if guide_list != None and control_guide not in guide_list:
        guide_list = adata.var_names.tolist() + [control_guide]
    elif guide_list == None:
        guide_list = adata.var_names.tolist()

    for guide in guide_list:
        pk_guide = adata[:, guide].X.flatten() / adata[:, guide].X.sum()
        entro[guide] = entropy(pk_guide, qk_guide)
    
    df = pd.DataFrame(entro, index=[result_field]).T.sort_values(by=result_field, ascending=False).head(n_top).reset_index()

    adata.uns[result_field] = df
    return

File: r_src/cluster_independent.R
library(parallel)
library(ggplot2)
library(dplyr)
library(transport)
library(entropy)

rank_by_relative_entropy <- function(
  adata, 
  control_guide = 'sgNon-targeting', 
  guide_list = NULL, 
  return_fig = FALSE
) {
  # 初始化熵存储列表
  entro <- list()
  
  # 计算 control_guide 的 qk 分布
  qk_guide <- colSums(adata[, control_guide]) / sum(adata[, control_guide])
  
  # 如果 guide_list 未提供，使用所有变量名称
  if (is.null(guide_list)) {
    guide_list <- colnames(adata)
  }
  
  # 计算每个 guide 的 pk 分布并计算 KL divergence
  for (guide in guide_list) {
    pk_guide <- colSums(adata[, guide]) / sum(adata[, guide])
    entro[[guide]] <- KL.plugin(pk_guide, qk_guide)
  }
  
  # 将熵数据转换为数据框，并排序
  df <- data.frame(Guide = names(entro), Shannon = unlist(entro)) %>%
    arrange(Shannon)
  
  # 绘制条形图
  p <- ggplot(df, aes(x = reorder(Guide, Shannon), y = Shannon, fill = Shannon)) +
    geom_bar(stat = "identity") +
    scale_fill_distiller(palette = "RdBu", direction = -1) +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    labs(title = "KL distance", x = "Guide", y = "Shannon Index")
  
  print(p)
  
  # 返回图形或数据框
  if (return_fig) {
    return(p)
  } else {
    return(df)
  }
}

rank_by_kernel_estimated_distance <- function(
  adata, 
  control_guide = 'sgNon-targeting', 
  guide_list = NULL, 
  n_permutation = 50, 
  n_process = 8, 
  return_fig = FALSE, 
  return_dataframe = TRUE, 
  sort_by_replicate = '_'
) {
  
  # 定义 KDE 计算函数
  compute_kde_statsmodels <- function(data) {
    kde <- density(data)
    return(kde$y)  # 返回密度值
  }
  
  # 定义工作者函数来计算实际和置换的 Wasserstein 距离
  kde_worker <- function(args) {
    guide <- args$guide
    control <- args$control
    spatial_coords <- args$spatial_coords
    data <- args$data
    n_permutation <- args$n_permutation
    
    if (guide == control) return(NULL)
    
    gene_a_expression <- as.numeric(data[, guide])
    gene_b_expression <- as.numeric(data[, control])
    
    # 将空间坐标和表达数据结合
    data_a <- cbind(spatial_coords[,1], spatial_coords[,2], gene_a_expression)
    data_b <- cbind(spatial_coords[,1], spatial_coords[,2], gene_b_expression)
    
    # 计算 KDE
    Za <- compute_kde_statsmodels(data_a)
    Zb <- compute_kde_statsmodels(data_b)
    
    # 计算 Wasserstein 距离
    actual_distance <- wasserstein1d(Za, Zb)
    
    combined_expression <- c(gene_a_expression, gene_b_expression)
    permuted_distances <- numeric(n_permutation)
    
    # 随机打乱数据并进行置换
    set.seed(42)
    for (i in 1:n_permutation) {
      permuted_expression <- sample(combined_expression)
      permuted_a <- permuted_expression[1:length(gene_a_expression)]
      permuted_b <- permuted_expression[(length(gene_a_expression)+1):length(permuted_expression)]
      
      permuted_Za <- compute_kde_statsmodels(cbind(spatial_coords[,1], spatial_coords[,2], permuted_a))
      permuted_Zb <- compute_kde_statsmodels(cbind(spatial_coords[,1], spatial_coords[,2], permuted_b))
      
      permuted_distances[i] <- wasserstein1d(permuted_Za, permuted_Zb)
    }
    
    # 计算 p-value
    p_value <- sum(permuted_distances >= actual_distance) / length(permuted_distances)
    
    return(c(guide, actual_distance, p_value))
  }
  
  # 如果 guide_list 未提供，使用所有变量
  if (is.null(guide_list)) {
    guide_list <- rownames(adata)
  }
  
  # 检查是否同时返回图像和数据框
  if (return_fig && return_dataframe) {
    stop('Error, can only return one of fig or dataframe!')
  }
  
  # 筛选数据
  filtered_data <- adata[rowSums(adata[, guide_list] > 0) > 0, ]
  spatial_coords <- as.matrix(filtered_data$spatial)
  
  # 使用 parallel 包的 mclapply 进行并行处理
  args_list <- lapply(guide_list, function(guide) list(guide = guide, control = control_guide, spatial_coords = spatial_coords, data = filtered_data, n_permutation = n_permutation))
  results <- mclapply(args_list, kde_worker, mc.cores = n_process)
  
  # 过滤结果并转换为数据框
  d_df <- do.call(rbind, results)
  d_df <- as.data.frame(d_df, stringsAsFactors = FALSE)
  colnames(d_df) <- c('guide', 'wd', 'p_value')
  d_df$wd <- as.numeric(d_df$wd)
  d_df$p_value <- as.numeric(d_df$p_value)
  
  d_df <- d_df %>% arrange(wd)
  
  # 可视化
  if (!is.null(sort_by_replicate)) {
    d_df$replicate <- sapply(strsplit(d_df$guide, sort_by_replicate), `[`, 1)
    positions <- seq(1, nrow(d_df)) * 2
    width <- 0.5
    
    p <- ggplot(d_df, aes(x = reorder(guide, -wd), y = wd)) +
      geom_bar(stat = "identity", width = width, fill = "steelblue") +
      theme_minimal() +
      theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
      labs(x = "Guide", y = "WD", title = "Kernel Estimated Distance")
    
    print(p)
    
  } else {
    p <- ggplot(d_df, aes(x = reorder(guide, wd), y = wd, fill = guide)) +
      geom_bar(stat = "identity") +
      theme_minimal() +
      theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
      labs(x = "Guide", y = "WD", title = "Kernel Estimated Distance")
    
    print(p)
  }
  
  if (return_dataframe) {
    return(d_df)
  } else if (return_fig) {
    return(p)
  } else {
    return(NULL)
  }
}

rank_by_proportion_within_threshold <- function(
  adata, 
  thresholds = seq(1, 501, length.out = 21),
  guide_list = NULL,
  control_guide = 'sgNon-targeting',
  method = 'bin',
  kernel_func = 'log',
  kernel_scoring = NULL,
  return_fig = FALSE
) {
  
  proportion <- list()
  
  # 如果 guide_list 没有提供，使用所有变量名称
  if (is.null(guide_list)) {
    guide_list <- rownames(adata)
  }
  
  # 二分法（bin）或者计数法（count）
  if (method == 'bin') {
    for (threshold in thresholds) {
      proportion[[as.character(threshold)]] <- c()
      for (guide in guide_list) {
        if (guide == control_guide) next
        
        guide_data <- adata[, adata[guide, ] > 0]
        ntc_data <- adata[, adata[control_guide, ] > 0]
        
        # 计算距离矩阵，并检查 guide 数据是否在 threshold 距离内
        in_distance <- sum(rowSums(as.matrix(dist(guide_data@obsm$spatial, ntc_data@obsm$spatial)) < threshold) > 0)
        proportion[[as.character(threshold)]] <- c(proportion[[as.character(threshold)]], in_distance / nrow(guide_data))
      }
    }
    
  } else if (method == 'count') {
    for (threshold in thresholds) {
      proportion[[as.character(threshold)]] <- c()
      for (guide in guide_list) {
        if (guide == control_guide) next
        
        guide_data <- adata[, adata[guide, ] > 0]
        ntc_data <- adata[, adata[control_guide, ] > 0]
        
        # 计算距离并对 guide 数据进行计数
        in_distance <- sum(guide_data[(rowSums(as.matrix(dist(guide_data@obsm$spatial, ntc_data@obsm$spatial)) < threshold)) > 0, guide])
        proportion[[as.character(threshold)]] <- c(proportion[[as.character(threshold)]], in_distance / nrow(guide_data))
      }
    }
    
  } else {
    stop('Error, method must be one of "bin" or "count"!')
  }
  
  # 将 proportion 转换为 data frame
  d_df <- melt(as.data.frame(proportion))
  d_df$variable <- rep(guide_list, each = length(thresholds))
  
  # 根据不同的 kernel 计算评分
  if (!is.null(kernel_scoring)) {
    sc <- apply(d_df, 1, function(x) sum(kernel_scoring * (1 - as.numeric(x))))
  } else {
    if (kernel_func == 'log') {
      ln_scoring <- log2(seq(1, exp2(10), length.out = length(thresholds)))
      sc <- apply(d_df, 1, function(x) sum(ln_scoring * (1 - as.numeric(x))))
    } else if (kernel_func == 'linear') {
      linear_scoring <- seq(0, 10, length.out = length(thresholds))
      sc <- apply(d_df, 1, function(x) sum(linear_scoring * (1 - as.numeric(x))))
    } else if (kernel_func == 'exp') {
      exp_scoring <- 2 ^ seq(1e-32, log2(10), length.out = length(thresholds))
      sc <- apply(d_df, 1, function(x) sum(exp_scoring * (1 - as.numeric(x))))
    }
  }
  
  # 创建 dataframe
  df <- data.frame(dist = sc, guide = guide_list)
  
  # 可视化
  p <- ggplot(df, aes(x = reorder(guide, dist), y = dist)) +
    geom_bar(stat = "identity", fill = "steelblue") +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    scale_y_log10() +
    labs(x = "Gene", y = "Distance score", title = "Distance Score by Guide")
  
  print(p)
  
  if (return_fig) {
    return(p)
  } else {
    return(NULL)
  }
}
